### DISCUSSION TITLE
Top sources to learn writing test cases

### DISCUSSION THREAD

**Discussion by haipinglu** [pykale/pykale#86] [Testing] [answered]
- High-level: [Testing fastai](https://fastai1.fast.ai/dev/test.html) is the best source I found so far that can guide us in testing. It shows a professional way to use **pytest** for pytorch-based libraries. Note its advice is more relevant to pytorch/DL, different from many online blogs/tutorials (some use mocks, which is not good as said below).

  Particularly, I think we should follow their recommendations in [Writing tests](https://fastai1.fast.ai/dev/test.html#writing-tests):
  - Avoid mocks; instead, think about how to create a test of the real functionality that runs quickly
  - Use module scope fixtures to run init code that can be shared amongst tests. When using fixtures, make sure the test doesn’t modify the global object it received, otherwise other tests will be impacted. If a given test modifies the global fixture object, it should either clone it or not use the fixture and create a fresh object instead.
  - Avoid pretrained models, since they have to be downloaded from the internet to run the test
  - Create some minimal data for your test, or use data already in repo’s data/ directory

- [A tutorial on writing tests for VAE - Pytorch](https://krokotsch.eu/cleancode/2020/08/11/Unit-Tests-for-Deep-Learning.html) but note that it uses unittest rather than PyTest
- Recommended pytest+pytorch examples to learn from [fastai1 tests](https://github.com/fastai/fastai1/tree/master/tests) and [Kornia tests](https://github.com/kornia/kornia/tree/master/test)
- Link to add on our docs/readme for [pytest in vscode](https://code.visualstudio.com/docs/python/testing) and [pytest in pycharm](https://www.jetbrains.com/help/pycharm/pytest.html).

**Comment by haipinglu** [comment]
@bobturneruk @RaivoKoot @XianyuanLiu @pz-white @sz144 For your information and comments please.

**Reply by RaivoKoot** [reply]
Ah amazing. This will help me a lot getting started.

**Comment by haipinglu** [comment, marked as answer]
Two important use cases
- Exceptions: https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions
- Fixtures: https://docs.pytest.org/en/stable/fixture.html#yield-fixtures-recommended
